ewatercycle/analysis/__init__.py

```diff
@@ -0,0 +1,136 @@
+import pandas as pd
+import os
+from typing import Union
+from hydrostats import metrics
+import matplotlib.pyplot as plt
+from matplotlib.dates import DateFormatter
+
+
+def hydrograph(
+    discharge: pd.DataFrame,
+    *,
+    reference: str,
+    precipitation: pd.DataFrame = None,
+    fname: Union[os.PathLike, str] = None,
+    dpi: int = None,
+    title: str = 'Hydrograph',
+    discharge_units: str = 'm$^3$ s$^{-1}$',
+    precipitation_units: str = 'mm day$^{-1}$',
+) -> plt.Axes:
+    """Plot a hydrograph.
+
+    This utility function makes it convenient to create a hydrograph from
+    a set of discharge data from a `pandas.DataFrame`. A column must be marked
+    as the reference, so that the agreement metrics can be calculated.
+
+    Optionally, the corresponding precipitation data can be plotted for
+    comparison.
+
+    Parameters
+    ----------
+    discharge : pd.DataFrame
+        Dataframe containing time series of discharge data to be plotted.
+    reference : str
+        Name of the reference data, must correspond to a column in the discharge
+        dataframe. Metrics are calculated between the reference column and each
+        of the other columns.
+    precipitation : pd.DataFrame, optional
+        Optional dataframe containing time series of precipitation data to be
+        plotted from the top of the hydrograph.
+    fname : str, Path
+        If specified, a copy of the plot will be saved to this path.
+    dpi : int, optional
+        DPI for the plot.
+    title : str, optional
+        Title of the hydrograph.
+    discharge_units : str, optional
+        Units for the discharge data.
+    precipitation_units : str, optional
+        Units for the precipitation data.
+
+    Returns
+    -------
+    ax : :obj:`matplotlib.axes.Axes`
+        Instance of a `~matplotlib.axes.Axes` object.
+    """
+    discharge_cols = discharge.columns.drop('reference')
+    y_obs = discharge[reference]
+    y_sim = discharge[discharge_cols]
+
+    fig, (ax, ax_tbl) = plt.subplots(
+        nrows=2,
+        ncols=1,
+        dpi=dpi,
+        figsize=(10, 10),
+        gridspec_kw={'height_ratios': [3, 1]},
+    )
+
+    ax.set_title(title)
+    ax.set_ylabel(f'Discharge ({discharge_units})')
+
+    y_obs.plot(ax=ax)
+    y_sim.plot(ax=ax)
+
+    handles, labels = ax.get_legend_handles_labels()
+
+    # Add precipitation as bar plot to the top if specified
+    if precipitation is not None:
+        ax_pr = ax.twinx()
+        ax_pr.invert_yaxis()
+        ax_pr.set_ylabel(f'Precipitation ({precipitation_units})')
+
+        for pr_label, pr_timeseries in precipitation.iteritems():
+            ax_pr.bar(
```

Comment:
Ideally, if precip data is from the same source as discharge data, it would be nice if they had the same color. Otherwise, maybe explicitly set a completely different color cycle, so there is no suggestion of a relationship between similarly colored discharge/precip.

Reply:
The first is really difficult to do, unless we explicitly pass some sort of mapping between the discharge/precipitation data. So I will re-use the color cycle from the discharge data to ensure the colors are different.

Reply:
Alternatively, we could pass a dict which maps the datasets to user-defined colors.

Reply:
Or, if the discharge and precip dataframes have the same column names, they could map to the same colors?

Reply:
I'm not sure if there is a simple way to do this in the current function. Any more tinkering and we may be better off rewriting the entire thing or implement it as a class. If you agree, I will make how to color the lines a new issue, so we can tackle it in a follow-up PR.

Reply:
Reusing the color cycle is acceptable for now 👍